File: gtci/03-sliding-window/cpp/3-longest-substring-with-k-distint-characters.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
   public:
    int findLength(const string &str, int k) {
        int maxLength = 0;

        int windowStart = 0;
        std::unordered_map<char, int> chars;

        for (int windowEnd = 0; windowEnd < str.size(); windowEnd++) {
            chars[str[windowEnd]] += 1;

            if (chars.size() == k)
                maxLength = std::max(maxLength, windowEnd - windowStart + 1);

            while (chars.size() > k) {
                chars[str[windowStart]] -= 1;

                if (chars[str[windowStart]] == 0) chars.erase(str[windowStart]);

                windowStart++;
            }
        }

        return maxLength;
    }
};

int main() {
    Solution sol;

    assert(sol.findLength("araaci", 2) == 4);
    assert(sol.findLength("araaci", 1) == 2);
    assert(sol.findLength("cbbebi", 3) == 5);

    std::cout << "All test cases passed." << std::endl;
    return 0;
}

File: gtci/03-sliding-window/python3/3-longest-substring-with-k-distint-characters.py
class Solution:
    def findLength(self, str1, k):
        max_length = 0

        window_start = 0
        chars = {}

        for window_end in range(len(str1)):
            if str1[window_end] not in chars:
                chars[str1[window_end]] = 0
            chars[str1[window_end]] += 1

            if len(chars) == k:
                max_length = max(max_length, window_end - window_start + 1)

            # if too many chars, shrink window until len(chars) <= k
            while len(chars) > k:
                chars[str1[window_start]] -= 1

                if chars[str1[window_start]] == 0:
                    del chars[str1[window_start]]

                window_start += 1

        return max_length


if __name__ == "__main__":
    sol = Solution()
    assert sol.findLength("araaci", 2) == 4
    assert sol.findLength("araaci", 1) == 2
    assert sol.findLength("cbbebi", 3) == 5
    print("All test cases passed.")
